aws/resource_aws_acm_certificate_test.go

```diff
@@ -0,0 +1,324 @@
+				),
+			},
+			// Test that we can change the tags
+			resource.TestStep{
+				Config: testAccAcmCertificateConfigWithChangedTags(domain, sanDomain),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAcmCertificateExists("aws_acm_certificate.cert", &conf, &tags),
+					testAccCheckAcmCertificateAttributes("aws_acm_certificate.cert", &conf, domain, sanDomain, "PENDING_VALIDATION"),
+
+					testAccCheckTagsACM(&tags.Tags, "Environment", "Test"),
+					testAccCheckTagsACM(&tags.Tags, "Foo", "Baz"),
+
+					resource.TestCheckResourceAttr("aws_acm_certificate.cert", "tags.%", "2"),
+					resource.TestCheckResourceAttr("aws_acm_certificate.cert", "tags.Environment", "Test"),
+					resource.TestCheckResourceAttr("aws_acm_certificate.cert", "tags.Foo", "Baz"),
+				),
+			},
+			// Test that validation times out if certificate can't be validated
+			resource.TestStep{
+				Config:      testAccAcmCertificateWithValidationConfig(domain, sanDomain),
+				ExpectError: regexp.MustCompile("Expected certificate to be issued but was in state PENDING_VALIDATION"),
+			},
+			// Test that validation fails if given validation_fqdns don't match
+			resource.TestStep{
+				Config:      testAccAcmCertificateWithValidationConfigAndWrongFQDN(domain, sanDomain),
+				ExpectError: regexp.MustCompile("Certificate needs .* to be set but only .* was passed to validation_record_fqdns"),
+			},
+			// Test that validation succeeds once we provide the right DNS validation records
+			resource.TestStep{
+				Config: testAccAcmCertificateWithValidationAndRecordsConfig(root_zone_domain, domain, sanDomain),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAcmCertificateExists("aws_acm_certificate.cert", &confAfterValidation, &tags),
+					testAccCheckAcmCertificateAttributes("aws_acm_certificate.cert", &confAfterValidation, domain, sanDomain, "ISSUED"),
+					testAccCheckAcmCertificateValidationAttributes("aws_acm_certificate_validation.cert", &confAfterValidation),
+				),
+			},
+			resource.TestStep{
+				ResourceName:      "aws_acm_certificate.cert",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func testAccAcmCertificateConfig(domain string, sanDomain string) string {
+	return fmt.Sprintf(`
+resource "aws_acm_certificate" "cert" {
+  domain_name = "%s"
+  validation_method = "DNS"
+  subject_alternative_names = ["%s"]
+
+  tags {
+    "Hello" = "World"
+    "Foo" = "Bar"
+  }
+}
+`, domain, sanDomain)
+}
+
+func testAccAcmCertificateConfigWithChangedTags(domain string, sanDomain string) string {
+	return fmt.Sprintf(`
+resource "aws_acm_certificate" "cert" {
+  domain_name = "%s"
+  validation_method = "DNS"
+  subject_alternative_names = ["%s"]
+
+  tags {
+    "Environment" = "Test"
+    "Foo" = "Baz"
+  }
+}
+`, domain, sanDomain)
+}
+
+func testAccAcmCertificateWithValidationConfig(domain string, sanDomain string) string {
+	return fmt.Sprintf(`
+resource "aws_acm_certificate" "cert" {
```

Comment:
Nitpick: this can be `%s` from `testAccAcmCertificateConfig()`

Reply:
Done.